### PR TITLE
async API

### PR DESCRIPTION
**Pull request by legokichi**
https://github.com/ramosbugs/oauth2-rs/issues/44#issuecomment-427739934

### CODE REVIEW

File: tests/lib.rs [hunk 1]
```diff
@@ -273,15 +276,18 @@ fn test_exchange_code_successful_with_complete_json_response() {
 fn test_exchange_client_credentials_with_basic_auth() {
     let mock = mock("POST", "/token")
         // base64(urlencode("aaa/;&") + ":" + urlencode("bbb/;&"))
-        .match_header("Authorization", "Basic YWFhJTJGJTNCJTI2OmJiYiUyRiUzQiUyNg==")
+        .match_header("Authorization", "basic YWFhJTJGJTNCJTI2OmJiYiUyRiUzQiUyNg==")
```
Comment (legokichi) [inline tests/lib.rs]: reqwest puts lowercase letters. 
https://docs.rs/reqwest/0.9.3/src/reqwest/async_impl/request.rs.html#171

**Comment by legokichi** [comment]
reqwest does not have HttpConnector API - 
 https://docs.rs/hyper/0.12.12/hyper/client/struct.HttpConnector.html

**Comment by ramosbugs** [comment]
thanks for the PR! I'm traveling this weekend, but I'll take a look next week.

File: src/lib.rs [hunk 2]
```diff
@@ -236,6 +237,24 @@ use prelude::*;
 
 const CONTENT_TYPE_JSON: &str = "application/json";
 
+#[derive(Clone)]
+/// This struct is used to maintain the strict URI encoding standard as proposed by RFC 3986
+struct StrictEncodeSet;
```
Comment (legokichi) [inline src/lib.rs]: https://github.com/rusoto/rusoto/blob/7640ca3de1e6486b0b25bce395810a9434a66ee5/rusoto/core/src/signature.rs#L593

**Comment by ramosbugs** [comment]
Sorry for the delay! Looking this PR over now.

> I'm thinking of using hyper instead of reqwest.
> reqwest does not have HttpConnector API -
> https://docs.rs/hyper/0.12.12/hyper/client/struct.HttpConnector.html

I think it's definitely useful (often required) for clients to be able to specify connection parameters, including TLS settings (trusted CA certs, cipher suites, etc.) and proxy configs. This functionality has been missing from this crate so far, so thank you for bringing it up!

However, it doesn't look like `hyper::client::HttpConnector` really lets us specify these parameters. Rather, that interface seems to deal with lower level network options. What sort of customization were you hoping to support via `HttpConnector`?

As an alternative, `reqwest::async::ClientBuilder` does seem to allow significant customization: https://docs.rs/reqwest/0.9.4/reqwest/async/struct.ClientBuilder.html. What if we add a `set_http_client` method to `oauth2::Client` so that users can pass in their own `reqwest::async::Client`s?

> - https://actix.rs/actix-web/actix_web/client/struct.ClientConnector.html#method.with_connector

it looks like this one refers to a different interface (`openssl`'s `SslConnector` struct rather than `hyper`'s `Connect` trait)

---

I'll submit comments on the rest of the PR soon.

File: src/lib.rs (hunk 2 above)
Comment (ramosbugs) [reply src/lib.rs]: I'm finding the chain of specs referenced by https://tools.ietf.org/html/rfc6749#appendix-B to be a bit hard to follow, so it's not super clear to me exactly which characters need to be escaped (especially since we're dealing with an HTTP header here, rather than a URI). This code does seem to maintain the current behavior as implemented in https://github.com/curl/curl/blob/master/lib/escape.c#L40, but I wonder if it would suffice to use  
[`USERINFO_ENCODE_SET`](https://docs.rs/percent-encoding/1.0.1/percent_encoding/struct.USERINFO_ENCODE_SET.html), which seems to be intended for username and password escaping (according to https://github.com/dtolnay/url/blob/master/UPGRADING.md).

@DeeUnderscore do you have any thoughts here, as the author of #41? (see usage below)

File: tests/lib.rs [hunk 3]
```diff
@@ -202,8 +203,9 @@ fn test_exchange_code_successful_with_minimal_json_response() {
             Url::parse(&(SERVER_URL.to_string() + "/token")).unwrap(),
         )),
     );
-    let token = client
-        .exchange_code(AuthorizationCode::new("ccc".to_string()))
+    let token = ::tokio::runtime::Runtime::new().unwrap().block_on(
```
Comment (ramosbugs) [inline tests/lib.rs]: since each test needs to do this, let's define a function that takes a future, creates a runtime, blocks on it, and returns the result. it's probably worth trying to minimize the test boilerplate.

File: src/lib.rs (hunk 2 above)
Comment (DeeUnderscore) [reply src/lib.rs]: By my reading, this adheres to `application/x-www-form-urlencoded`—everything outside of the permitted bytes is encoded. [This WHATWG spec](https://url.spec.whatwg.org/#application/x-www-form-urlencoded) seems to be essentially equivalent to what RFC6749 wanted when it described the encoding in Appendix B. 

URLs can use forms of percent encoding which aren't `application/x-www-form-urlencoded`, since they permit more byte ranges (presumably to make URLs more readable). `USERINFO_ENCODE_SET` ([WHATWG ref](https://url.spec.whatwg.org/#userinfo-percent-encode-set)), in particular, is intended for the userinfo component of a URL, ie. the `user:password` in `https://user:password@example.org/`.

In practical terms, you could probably get away with a less strict encoding, and I suspect most OAuth implementations in the wild avoid using characters which could cause problems anyway. But, for closest adherence to the standard I'd recommend what this PR is doing.

Comment (ramosbugs) [reply src/lib.rs]: Thanks for the reference and rationale! That makes sense; I agree with the recommendation.

The encoding here does seem to deviate ever so slightly from the WHATWG spec, though:
- spaces encoded with `%20` instead of `+`
- asterisks (`0x2A`) excluded from the encoding set but mentioned by WHATWG
- tildes (`0x7E`) included in encoding set but excluded by WHATWG

None of these deviations seems particularly problematic, though it's probably safer for interoperability to err on the side of percent-encoding more values rather than fewer. Unless there's an error in the WHATWG spec, we should probably remove tildes from this set.

File: src/lib.rs [hunk 4]
```diff
@@ -807,7 +828,7 @@ impl<EF: ExtraTokenFields, TT: TokenType, TE: ErrorResponseType> Client<EF, TT,
     pub fn exchange_code(
         &self,
         code: AuthorizationCode
-    ) -> Result<TokenResponse<EF, TT>, RequestTokenError<TE>> {
+    ) -> impl Future<Item = TokenResponse<EF, TT>, Error = RequestTokenError<TE>> + Send + 'static {
```
[the remark this answers is not in the document]
Comment (legokichi) [reply src/lib.rs]: tokio requires [`send + 'static`](https://tokio.rs/blog/2018-03-tokio-runtime/) because it adopts the [work-stealing algorithm](https://tokio-rs.github.io/tokio/tokio/runtime/fn.run.html).

File: src/lib.rs [hunk 5]
```diff
@@ -808,6 +841,14 @@ impl<EF: ExtraTokenFields, TT: TokenType, TE: ErrorResponseType> Client<EF, TT,
         &self,
         code: AuthorizationCode
     ) -> Result<TokenResponse<EF, TT>, RequestTokenError<TE>> {
+        block_on(self.exchange_code_async(code))
+    }
+
+    /// async edition
+    pub fn exchange_code_async(
```
Comment (legokichi) [inline src/lib.rs]: futures trait also [changes greatly](https://rust-lang-nursery.github.io/futures-rs/) in [Rust 2018](https://rust-lang-nursery.github.io/edition-guide/rust-2018/control-flow/async-await-for-easier-concurrency.html).
How about doing this as a provisional response for the past few months?

* exchange_code_async
* exchange_password_async
* exchange_client_credentials_async
* exchange_refresh_token_async

**Comment by legokichi** [comment]
Here are the implementations of https_connector.

* native-tls: https://github.com/hyperium/hyper-tls
* rustls: https://github.com/ctz/hyper-rustls

File: src/lib.rs (hunk 5 above)
Comment (legokichi) [reply src/lib.rs]: Alternative choice: just constructs and returns [`http::request::Request`](https://docs.rs/http/0.1.13/http/request/struct.Request.html) object.
User can chose own http request library e.g. curl, reqwest, hyper, actix-web ...etc.

File: src/lib.rs [hunk 6]
```diff
@@ -211,31 +211,62 @@
...
+    #[inline]
+    fn contains(&self, byte: u8) -> bool {
+        let upper = byte >= 0x41 && byte <= 0x5a;
+        let lower = byte >= 0x61 && byte <= 0x7a;
+        let numeric = byte >= 0x30 && byte <= 0x39;
+        let hyphen = byte == 0x2d;
+        let underscore = byte == 0x5f;
+        let tilde = byte == 0x7e;
+        let period = byte == 0x2e;
+        !(upper || lower || numeric || hyphen || underscore || tilde || period)
```
Comment (ramosbugs) [inline src/lib.rs]: based on the conversation above, I think we should probably remove `tilde`

File: src/lib.rs (hunk 5 above)
Comment (ramosbugs) [reply src/lib.rs]: I think it's fine to put the pre-Rust 2018 async functionality into this crate. However, I think we should do this via a separate `Client` implementation (i.e., `oauth2::async::Client`). Then, when we're ready to add support for Rust 2018 async, we can call that `oauth2::async2018::Client` or something. Each of these can present the same `exchange_*` functions, just with different return types. This change will probably require some refactoring to share as much code as possible between the different `Client` implementations.

**Comment by ramosbugs** [comment]
> Here are the implementations of https_connector.
> 
> * native-tls: https://github.com/hyperium/hyper-tls
> * rustls: https://github.com/ctz/hyper-rustls
> * https://github.com/pimeys/hyper-alpn
> * https://github.com/sfackler/hyper-native-tls
> * https://github.com/sfackler/hyper-openssl

Ohh, I see. So it looks like what we really want to accept is a struct that implements the `hyper::client::connect::Connect` trait, which can do anything it wants as long as it yields a `(hyper::client::connect::Connect::Transport, hyper::client::connect::Connected)`.  Something like what you mentioned above should work, e.g.:
```
pub struct Client<C: Connect, EF: ExtraTokenFields, TT: TokenType, TE: ErrorResponseType> {
    connector: C,
    ...
}

// new() sets a sensible default (and secure) connector (require HTTPS, etc.)

pub fn with_connector(mut self, connector: C) -> Self {
    self.connector = connector;

    self
}
```

**Comment by bachp** [comment]
I would like to use this in the context of actix-web. So I would be very happy if there is a way to specify the client library to use (reqwest, actix, ...). Did you already give any toughts on how this could be done?

If it would be desired to have this as a feature I might be able to come up with a PR that allows to select actix-web as a replacement for reqwest.

**Comment by ramosbugs** [comment]
> I would like to use this in the context of actix-web. So I would be very happy if there is a way to specify the client library to use (reqwest, actix, ...). Did you already give any toughts on how this could be done?

I think the simplest approach to supporting multiple HTTP clients is probably via a trait. Since the `hyper::client::connect::Connect` trait is quite low level, I think it's a good candidate. This also lets people use `hyper` by default, which seems to have a significantly larger volume of downloads on crates.io than `reqwest` or `actix-web`. Other client libraries could potentially be used by writing wrappers that implement the `hyper::client::connect::Connect` trait on top of those libraries.

Alternatively, we could probably do something at compile time based on feature flags, but I haven't looked into that much, and it feels like that would be harder to maintain over time.

Was there specific functionality from `actix-web` that you were looking to use in conjunction with this crate? Understanding the use case a bit more might help illuminate the right solution.

I think it's also probably worth filing a separate Github issue for this so that we can refocus #44 on the async discussion, and the new issue on the need for a generic HTTP client interface.

**Comment by ramosbugs** [comment]
latest master now features an async API with pluggable HTTP clients
